hack/BUILD.bazel

```diff
@@ -9,3 +11,21 @@ exports_files(glob([
     "boilerplate/*.txt",
     "*.sh",
 ]))
+
+go_library(
+    name = "go_default_library",
+    srcs = ["tools.go"],
+    importpath = "sigs.k8s.io/cluster-api-provider-aws/hack",
+    deps = [
+        "//vendor/k8s.io/code-generator/cmd/client-gen:go_default_library",
+        "//vendor/k8s.io/code-generator/cmd/conversion-gen:go_default_library",
+        "//vendor/k8s.io/code-generator/cmd/deepcopy-gen:go_default_library",
+        "//vendor/k8s.io/code-generator/cmd/defaulter-gen:go_default_library",
+        "//vendor/k8s.io/code-generator/cmd/informer-gen:go_default_library",
+        "//vendor/k8s.io/code-generator/cmd/lister-gen:go_default_library",
+        "//vendor/k8s.io/code-generator/cmd/register-gen:go_default_library",
+        "//vendor/k8s.io/code-generator/cmd/set-gen:go_default_library",
+        "//vendor/sigs.k8s.io/controller-tools/cmd/controller-gen:go_default_library",
```

Comment:
This is out of scope for this PR but I'm interested in what others think.

The benefits to keeping source code in tree as I see them are: being able to run the commands across platforms and updating versions of these tools is as easy as checking in a new version to vendor/. Folks developing will pull in changes and have the new version, no effort on their part. What else am I missing?

The downsides as I see it are that go mod doesn't support vendoring code that isn't directly referenced and we end up with `hack/tools.go` + the vendor copy, which will copy in the entire project and not just the necessary files.

So my question: Why are we treating controller-tools/code-generator in a special way from any other external dependency?

Reply:
One reason for keeping the vendor directory around is that we don't have to worry about dependencies being renamed (case sensitivity), moved, or referenced tags/hashes no longer being present.

Reply:
regarding controller-tools and other utilities referenced from `hack/tools.go`, we ideally want to use the versions that are compatible with the versions of controller-runtime and controller-tools that we are using. If we don't leverage the vendored versions, then we have to manually keep those in sync another way. That said, with the way that gomod handles dependencies, we may not be ensuring that they are compatible versions today :(

Reply:
I recognize things change as a theoretical problem but in practice I have never found it to be a problem. It would be good to understand the decision made to keep vendor and what problems we're solving for.

Reply:
Since we have most dependencies in bazel, could we use bazel to keep these dependencies in sync?

Reply:
@chuckha I think it's likely to be more of an issue at the current point in time due to changes made between controller-runtime 1.x and the 2.x betas we are currently relying on.

I'd be fine with defering to Bazel to manage the dependencies for the generation tooling, we'd just need to document keeping those dependencies updated along with controller-tools.

Reply:
ok, I'm not saying we should, I'm mostly thinking about this from the problems I'm about to slam my head into with the bootstrap controller provider. I really appreciate the discussion, thank you